RFP-01.md

```diff
@@ -0,0 +1,169 @@
+
+### Deliverables
+
+- Contract system for `SubPoolFactory`, `SubPool`, and `SignedSettlement`
+- Comprehensive documentation
+- Open-source code repository
+
+### Interface
+
+```solidity
+interface Auth {
+    /**
+     * Allows `usr` to perform root-level functions
+     */
+    function rely(address usr) external auth;
+    /**
+     * Denies `usr` from being able to perform root-level functions
+     */
+    function deny(address usr) external auth;
+    /**
+     * Determine if `usr` is able to perform root-level functions
+     */
+    function wards(address usr) external view returns (bool);
+}
+
+interface SubPool is Auth {
+    /**
+     * Determine the amount required to `heal` the pool
+     * @return amt of immutable collateralToken
+     * @return cowAmt of COW
+     */
+    function dues() external view returns (uint256 amt, uint256 cowAmt);
+    /**
+     * Pull in from `msg.sender` the dues required to `heal` the pool
+     */
+    function heal() external;
+}
+
+interface SubPoolFactory is Auth {
+    // --- view functions ---
+
+    /**
+     * View function for determining the minimum collateral required if using `token`
+     * @param token The token to retrieve the collateral requirement for
+     */
+    function collateral(IERC20 token) external view returns (uint256);
+    /**
+     * View function for determining if a given solver EOA has a correctly collateralized
+     * sub-pool and has not announced an exit.
+     * @param subPoolSolver The solver to check
+     * @return bool if the solver is allowed to solve
+     */
+    function canSolve(address subPoolSolver) external view returns (bool);
+    /**
+     * View function for determining the backend URI for the sub-pool
+     * @return string URI
+     */
+    function backendUri(SubPool pool) external view returns (string memory);
+
+    // --- permissionless functions ---
+    /**
+     * Deploy a `SubPool` deterministically based on `msg.sender`.
+     * @notice The `msg.sender` SHOULD NOT be a contract.
+     * @param token The nominated token to use for collateral
+     * @param amt How much collateral to add to the sub-pool (must be minimally collateralised)
+     * @param cowAmt Amount of COW to add as collateral
+     * @param backendUri The URI to send the batches to the solver
+     * @dev Emits an event `SubPoolCreated` when creating a sub-pool
+     */
+    function create(IERC20 token, uint256 amt, uint256 cowAmt, string calldata backendUri) external;
+
+    // --- cow dao bonding pool authed functions ---
+
+    /**
+     * Set the delay for a sub-pool to exit
+     * @dev Emits an event when setting a delay
+     * @param delay The delay in seconds
+     */
+    function setExitDelay(uint256 delay) external auth;
+    /**
+     * Allow `token` to be used as collateral for the sub-pool.
+     * @dev Emits an event `AllowCollateral` when allowing/updating a token
+     */
+    function allowCollateral(IERC20 token, uint256 min) external auth;
```

Comment:
If there are multiple collaterals do all need to be provided or some (maybe document)?

Reply:
> If there are multiple collaterals do all need to be provided or some (maybe document)?

By multiple collaterals, do you refer to:
i. The a subpool factory **owner** (such as CoW DAO bonding pool) allowing subpools with different collaterals, but where each subpool itself consists only of collateral 'A' + COW; or
ii .The subpool itself may be collateralised with tokens X + Y + Z + COW?

Reply:
I think more importantly, we need to have some idea of how to handle the ETH from the MEV Blocker rebates (this adds a third collateral potentially). @fleupold how did you envision that this would work?

Also, particularly as well, what should happen in the case that the solver is building from the small bond to the target bond, and they choose to withdraw prior to the 365 day period ending?

Reply:
This does seem important. How important is the 365 day period? And what do we want in the contracts related to that? For handling eth, seems like there are two options here, at exit solver gets their collateral token, cow token and eth back or they just get the collateral and cow tokens back without the eth. In that case we should also have some config regarding where the eth should be sent to?

Reply:
> How important is the 365 day period?

I would not enshrine the 365 day period in the contracts anywhere (I think the signer can simply seize to provide signatures to solvers that have not reached their target deposit after a certain time). If a solver wants to withdraw they should be able to withdraw any token balance they hold after the challenge period.

As to how to deal with the MEV Blocker ETH kickbacks, I see two options:
1. Add ETH as a collateral type that can be used instead of yield bearing USD
2. Have the refund recipient be a smart order that automatically wraps and swaps ETH for yield bearing USD with `receiver` being the solvers pool (using e.g. a `TradeAboveThreshold` smart order with pre-interactions)

Maybe we should even get rid of `collateral`, `allowCollateral`, etc. and have the collateral requirements be communicated off-chain as well? Otherwise we would definitely need a more sophisticated structure here (e.g. you need this much CoW and one of the following (50k USDC, USDT or 50 ETH).

Reply:
> I would not enshrine the 365 day period in the contracts anywhere (I think the signer can simply seize to provide signatures to solvers that have not reached their target deposit after a certain time). If a solver wants to withdraw they should be able to withdraw any token balance they hold after the challenge period.

This lacks certainty as to what happens to funds in the case of:

1. Withdrawal prior to the end of the 365 day period - what happens to accumulated funds over the minimum?
2. What happens after the 365 day period and the solver has failed to reach the target?

The 365 days is enshrined in CIP-44, and the above questions are very reasonable for a prospective solver to ask before even committing funds to start with. 

I do fear that by moving more and more elements off-chain will present an onerous administrative burden and that we should seize this initiative for true permissionless (save bond) access to the batch auction.

Reply:
Sorry for not being clear

> 1. Withdrawal prior to the end of the 365 day period - what happens to accumulated funds over the minimum?

They go back to the solver

> What happens after the 365 day period and the solver has failed to reach the target?

Whatever is providing signatures for the solver to use the settlement wrapper contract, seizes to do so leaving the solver with the choice to leave their capital idle in the pool or withdraw eventually.

--- 
I fear that by putting all the logic we know of right now on chain (which inevitably will be a subset of the logic the bonding pool will enforce in 6 months as requirements change) we will end up with a hybrid system where whatever off-chain component is handing out signatures is doing some check off-chain and some checks on-chain (and it won't necessarily be clear to the solver why they are not receiving signatures from on-chain state only and even if there is no guarantee the signer cannot start withholding signatures for a solver it no longer likes). For complete permissionless-ness, solvers need to create a full bonding pool.

However, since signature under-writing has to happen off-chain I think it makes sense to say "the component that underwrites has certain criteria that it publishes somewhere publicly" and that's the source of truth. IMO it's just a waste of time and complexity to enshrine things like exact collateral requirements, timelines on when those should be reached, etc. into this wrapper contract.

Imo the main goal of the Subpool contract is to replace the Safes we currently use to store collateral with an onchain system that has better custody (ie a way to somewhat permissionless-ly withdraw) and allows solvers to adjust frequently changing aspects about their pool (e.g. submission address, backend url) without intervention from our side. But really the more important contribution here is the wrapper contract...

Reply:
OK, now this follows a `bill` and `dues` methodology and `canSolve` will be `false` if there is an outstanding `bill`, or the solver has announced and exit. This allows for arbitrary collateralisation (with the mandate of a minimum amount of `cow` for spam prevention). The rest is off-chain.